File: Exercise_2.cpp
// Time Complexity                              :  O(4^N). Since at each step we have 4 options.
// Space Complexity                             :  O(2*N). O(N) due to the recursion stack and O(N) due to creating a string in each recursion.
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/expression-add-operators/solution/

class Solution {
private:
    int n;
    vector<string> ans;
public:
    vector<string> addOperators(string num, int target) {
        n = num.size();
        dfs("", num, target, 0, 0, 0);
        return ans;
    }
    
    // recursion
    void dfs(string expr, string num, int target, int idx, long val, long tail) {
        if(idx == n) {
            if(val == target) {
                ans.push_back(expr);
                return;
            }
        }
        
        int len = 1;
        for(int i=idx;i<n;i++) {
            if(i != idx and num[idx] == '0') continue;
            string str = num.substr(idx, len++);
            long curr = stol(str);
            if(idx == 0) {
                dfs(expr + str, num, target, i+1, curr, curr);
            } else {
                // +
                dfs(expr + '+' + str, num, target, i+1, (val - tail) + (tail + curr), curr);
                // -
                dfs(expr + '-' + str, num, target, i+1, (val - tail) + (tail - curr), -curr);
                // *
                dfs(expr + '*' + str, num, target, i+1, (val - tail) + (tail * curr), (tail*curr));
            }
        }
        
    }
};

File: Exercise_2.py
# Time Complexity: O(n*2^n)
# Space Complexity: O(n)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this : No


class Solution(object):
    def addOperators(self, num, target):
        """
        :type num: str
        :type target: int
        :rtype: List[str]
        """
        result = []
        
        if num == None or len(num) == 0:
            return result
        
        def helper(index, path, calc, tail):
            #base
            if index==len(num):
                if calc==target:
                    result.append(path)
                return
            
            #logic
            for i in range(index, len(num)):
                if index!=i and num[index]=='0':
                    continue
                curr = int(num[index:i+1])
                if index==0:
                    helper(i+1, path+str(curr), curr, curr)
                else:
                    helper(i+1, path+ '+' +str(curr), calc+curr, curr)
                    helper(i+1, path+ '-' +str(curr), calc-curr, (-1)*curr)
                    helper(i+1, path+ '*' +str(curr), calc-tail+tail*curr, tail*curr)
        
        helper(0, '', 0, 0)
        
        return result
                    